[1단계 - 자동차 경주] 이지윤 미션 제출합니다.

dd-jiyun opened the pull request:
안녕하세요 주노 🙌 
이번 미션도 잘 부탁드립니다!

## 🤔 고민했던 부분
### 자동차는 움직일 수 있다?
처음에는 자동차가 '움직일 수 있다'는 가능성에만 초점을 두고, 움직였는지를 true/false로 판단했습니다.
그런데 자동차가 실제로 '움직였다'는 것을 판단하려면, 그 `전후의 위치가 달라야 하지 않을까?` 라는 생각이 들었습니다!
움직임이 위치의 변화로 표현되는 것이 자연스럽다고 생각해서 자동차가 이름과 위치를 상태로 갖게 하였습니다.

### 랜덤 값을 어떻게 테스트 하지?
처음에는 `canMove()` 메소드 안에서 직접 `random.nextInt()`를 호출해서 랜덤값이 4 이상이면 움직이도록 구현했습니다.
하지만, 이 방법을 이용하니 테스트할 때마다 랜덤한 값이 생성되어서 매번 결과가 달라지는 문제가 생겼습니다.
그래서 랜덤 값을 외부에서 주입받도록 수정하니 `canMove()`는 랜덤 값이 들어오는지, 고정된 값이 들어오는지 알 필요 없이 `들어온 값이 4 이상의 값인지`를 확인할 수 있고 테스트에서는 고정된 값을 넣어 원하는 결과를 반환할 수 있게 하였습니다!

- 이전
```java
public void canMove() {
    if (random.nextInt(10) > 3) {
        position++;
    }
}
```
- 이후
```java
public void canMove(int randomValue) {
    if (randomValue > 3) {
        position++;
    }
}

```

## 🙌 궁금한 점
- 현재 저의 코드에서는 자동차가 위치를 상태로 가지고 `canMove()` 메소드에서 움직일 수 있는지 판단하는 두 책임을 가지고 있어요. SRP를 준수하기 위해서는 이 책임을 분리해야 하는 건가요? 
- 이번에 기능이 많지 않지만, 커밋을 엄청 작게 했어요. 주노는 어떤 것을 기준을 두고 커밋을 올리는 지 궁금해요!!

dd-jiyun committed:
docs: 기능 구현 목록 작성

dd-jiyun committed:
feat: 자동차 생성

dd-jiyun committed:
test: 자동차 생성 테스트

dd-jiyun committed:
docs: 자동차 생성 구현

dd-jiyun committed:
feat: 자동차 이동

dd-jiyun committed:
test: 자동차 이동 테스트

dd-jiyun committed:
docs: 자동차 이동 구현

Choi-JJunho commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,45 @@
+class CarTest {
+
+    @Test
+    @DisplayName("자동차는 정상적으로 이름을 갖는다.")
+    void shouldReturnNameForCar() {
+        // given
+        Car car = new Car("jiyun");
+
+        // when & then
+        assertThat(car.getName()).isEqualTo("jiyun");
+    }
```
자동차가 이름을 갖지 않는 경우에는 어떤 상황으로 판단해야한다고 생각하시나요?
(예외처리?, 정상케이스?)

dd-jiyun replied:
요구 사항에 `자동차는 이름을 갖는다`를 보니 이름이 없으면 안될 거 같아요.
null이거나 빈 값을 입력했을 경우에 예외를 반환해서 정상적인 입력을 유도하는 게 좋을 거 같습니다! 😄
감사합니다 !

Choi-JJunho commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public void canMove(int randomValue) {
+        if (randomValue > 3) {
```
3이라는 숫자를 의미있는 상수로 분리해보는건 어떨까요?

키워드: Magic number

dd-jiyun replied:
매직 넘버에 대해 알려주셔서 감사합니다 👍🏻 

알아보니 ‘의미 있는 이름으로 대체할 수 있는 숫자’를 말한다는 걸 알게 되었고, 그 의미를 고려했을 때 3처럼 특정한 의미를 갖는 숫자는 하드 코딩보다는 상수로 분리하는 게 더 좋겠다는 생각이 들었습니다 👍🏻

추가로 제 코드에 position을 0으로 초기화하는 부분이 고민이 되더라고요.
‘초기화’라는 문맥에서는 0이 워낙 자연스럽게 사용되어서 `이 경우에도 상수로 분리해야 할까?` 하는 의문이 생겼습니다.
저는 ‘딱 봐서 의미가 명확하지 않은 숫자’일 때 상수로 분리하는 게 좋다고 생각했거든요!

혹시 이런 경우에도 상수로 분리하는 걸 추천하시는지 궁금합니다 😆

Choi-JJunho replied:
> 딱 봐서 의미가 명확하지 않은 숫자’일 때 상수로 분리

저도 지윤님의 이 의견에 동감합니다~! 😄
지금처럼 position을 초기화하는 부분에서는 상수로 분리하지 않아도 괜찮을 것 같다는 생각이에요!

Choi-JJunho commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,23 @@
+        return position;
+    }
+
+    public void canMove(int randomValue) {
+        if (randomValue > 3) {
+            position++;
+        }
+    }
```
> 현재 저의 코드에서는 자동차가 위치를 상태로 가지고 canMove() 메소드에서 움직일 수 있는지 판단하는 두 책임을 가지고 있어요. SRP를 준수하기 위해서는 이 책임을 분리해야 하는 건가요?

저같은 경우에는 메소드명과 파라미터명으로 표시할 수 있지 않을까 생각이 들어요
지윤님이 작성해주신 `canMove`라는 메소드 명은 boolean을 반환할 것 같은 네이밍을 가지고 있다고 생각해요

isXXX, canXXX 과 같은 `가능 여부`를 추론할만한 메소드명이 아닌 다른 메소드명으로 대체해볼 수 있다면 어떤 네이밍이 있을까요?

또한 파라미터인 randomValue도 Car 클래스 내부에서만 보기에는 랜덤한 값 그 이상의 의미를 가지지 못하는 네이밍이 된다고 생각해요. randomValue는 어떤 역할을 하는지 생각해보고 적절한 네이밍이 어떤것이 있을지 한번 고민해볼까요? 🚗

dd-jiyun replied:
주노가 말씀해주신 대로 `canMove`라는 이름이 boolean을 반환할 것 같은 인상을 주네요!
`canMove` 메소드가 '움직일 수 있다면 움직인다'는 역할을 하기 때문에 네이밍을 다시 고민해보게 되었어요.

- movable : 의미는 "움직일 수 있는"이라는 뜻이라 `canMove`처럼 조건을 확인하는 메서드명처럼 느껴져요.
- moveForward : 전진을 시도하는 메서드처럼 보여서 괜찮아 보이긴 하지만, 조건에 따라 실제로 움직이지 않을 수도 있어서 이 이름이 적절할지 조금 고민이 돼요.

그래도 움직임을 명확하게 표현할 수 있는 `moveForward`을 사용하는 것으로 결정했습니다! 

> 또한 파라미터인 randomValue도 Car 클래스 내부에서만 보기에는 랜덤한 값 그 이상의 의미를 가지지 못하는 네이밍이 된다고 생각해요. randomValue는 어떤 역할을 하는지 생각해보고 적절한 네이밍이 어떤것이 있을지 한번 고민해볼까요? 🚗

말씀해주신 내용을 생각해보니 움직임을 담당하는 메소드의 파라미터가 랜덤이라는 것을 몰라도 되는 것 같네요!
그래서 다른 표현을 생각해봤어요. 

자동차의 엑셀을 밟는 것처럼 일정 힘 이상의 값이 들어오면 움직이고, 그렇지 않으면 움직이지 않는 것을 떠올려서 `power`라는 파라미터 명으로 변경했습니다. 그래서 움직임을 담당하는 메소드는 더 이상 랜덤이라는 것을 알지 못하게 되었습니다!

Choi-JJunho replied:
좋은 네이밍이네요! 👏👏 
객체지향적으로 개발한다는 말은 우리가 살고있는 현실세계를 객체 중심으로 모델링한다는 말이 되기도 해요

아마 읽어보셨을 수도 있겠지만 [조영호님의 객체지향의 사실과 오해](https://product.kyobobook.co.kr/detail/S000001628109)라는 책을 추천드려봅니다 :)

Choi-JJunho commented on README.md:
```diff
@@ -0,0 +1,9 @@
+## 🚗 자동차 경주
```
> 이번에 기능이 많지 않지만, 커밋을 엄청 작게 했어요. 주노는 어떤 것을 기준을 두고 커밋을 올리는 지 궁금해요!!

커밋을 아주 잘 작성해주셨어요! 👏
저는 기능별로 쪼개서 커밋하곤 합니다 👍

TDD를 하는 상황을 가정한다면 테스트가 성공하는 단위가 될 것 같네요 😄

Choi-JJunho reviewed (commented):
아주 빠르게 1단계를 제출해주셨네요! 👏 
궁금한점과 고민해볼만한 부분에 대해 코멘트 남겼습니다~!
확인 부탁드릴게요 👍

dd-jiyun committed:
refactor: 자동차 이름 검증 추가

- 자동차 이름에 null이나 빈 값이 입력될 경우 예외가 발생합니다.

Choi-JJunho commented on src/main/java/RandomGenerator.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.Random;
+
+public class RandomGenerator {
+    Random random = new Random();
+
+    public int getRandomValue() {
+        return random.nextInt(10);
+    }
+}
```
다른 분들 리뷰 달다가 생각나서 추가 코멘트를 남기러 왔습니다!

만약 랜덤에 대한 요구사항이 다음과 같이 변경되면 어떻게 될까요?
예상되는 변경 지점을 이야기해보고 변경에 유연하게 대응하기 위한 구조는 어떤 구조가 있을지 생각해볼까요?

> 상황 A) 랜덤한 숫자 중 400 이하의 숫자인 경우 자동차는 움직인다.

> 상황 B) 랜덤한 숫자 중 3으로 나눈 나머지가 1인 경우 자동차는 움직인다.

dd-jiyun replied:
지금 저의 코드에서는 `n보다 커야 움직일 수 있다`라는 규칙에만 유효한 것 같아요.
말씀해주신 상황 B처럼 규칙이 바뀐다면, 자동차 클래스의 내부 로직이 변경되어야 해서 유연하게 대응하기 어려운 것 같아요.

그래서 아래 두 가지 구조를 생각해봤어요.

- 첫 번째는 움직임의 규칙을 인터페이스로 분리하는 방식이에요.
자동차가 스스로 판단하고 움직일 수 있는 자율적인 객체가 되고, 규칙이 바뀌어도 인터페이스만 교체할 수 있다고 생각했어요.

- 두 번째는 자동차는 단지 움직이는 행위만을 책임지고, 외부에서 "움직여라"는 명령을 내리는 방식이에요.
그럼 자동차는 규칙을 전혀 알지 못하기 때문에 영향을 받지 않게 되는 거죠.

위의 두 가지를 관점의 차이로 보면, 능동적인 객체와 수동적인 객체로 나뉠 수 있을 거 같아요.
여기서 `자동차의 책임은 어디까지 일까?`에 대해서 고민이 생겼습니다. 🤔

도메인 관점으로 보고 판단까지 책임지는 것과 행동만 명확하게 책임지는 것 중 어떤 것이 좋은 설계인지 대해 주노의 의견이 궁금합니다!

Choi-JJunho replied:
좋은 고민이네요!
인터페이스 분리방식을 생각하고 드렸던 질문이였지만 능동적인 객체와 수동적인 객체에 대한 고민까지 생각하시다니 🤔 
제가 더 배워가네요 ㅎㅎㅎ

> 움직임의 규칙을 인터페이스로 분리하는 방식이에요

이 방식에 대해서는 잘 알고 계신 것 같아 관련한 키워드만 남겨드리고 넘어가볼게요! `전략패턴` 이라는 디자인패턴에 대해 알아보면 좋을 것 같아요 :)

> 자동차는 단지 움직이는 행위만을 책임지고, 외부에서 "움직여라"는 명령을 내리는 방식

해당 방식을 차용하면 자동차 자체가 외부 규칙(랜덤한 숫자, 4이상 등)을 몰라도 되기 때문에 격리된 상황에서 존재할 수 있어서 순수한 객체가 될 수 있다는 점이 좋게 다가오는 것 같아요

이 경우에는 Car 객체와 규칙을 하나로 가지는 어떤 객체를 만드는 식으로 composition(합성)을 사용해볼 수도 있을 것 같네요

```java
public interface MoveStrategy {
    boolean canMove();
}

public class MovableCar {
    private final Car car;
    private final MoveStrategy strategy;

    public MovableCar(Car car, MoveStrategy strategy) {
        this.car = car;
        this.strategy = strategy;
    }

    public void tryMove() {
        if (strategy.canMove()) {
            car.move();
        }
    }

    public int getPosition() {
        return car.getPosition();
    }
}
```
약간 이런 구조가 되려나요? 😄 

이 경우 책임분리는 명확해서 좋지만 다른 사람들이 Car 객체를 사용할 때 지켜져야하는 요구사항을 무시할 수 있게된다는 점은 아무래도 경계해볼만 한 부분이라고 생각이드네요 🤔 

정답이 있는 부분은 아니라서 어떤것이 옳다! 라고는 말씀드리긴 어려운 부분이네요!
다만 저의 개인적인 생각은 `지켜져야하는 정책이 있다!` 라는 사실을 Car 객체가 아는것이 더 안전한 객체를 만들 수 있지 않을까 생각이 들어요 (어디까지나 저의 주관적인 생각입니다!!)

dd-jiyun committed:
refactor: 자동차 객체가 자세한 움직임 규칙을 알지 못하게 수정

- 자동차 객체의 움직임 메서드 및 매개변수 이름을 수정합니다.

dd-jiyun committed:
refactor: final 키워드 추가

- 파라미터와 필드의 불변을 위해 final 키워드를 추가합니다.

Choi-JJunho commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    public void moveForward(final int power) {
+        if (MIN_POWER_TO_MOVE < power) {
+            position++;
+        }
+    }
```
power라는 명칭 좋네요! 자동차라는 객체와 잘 어울린다고 생각해요

별도의 이야기로 저는 자동차미션을 구현했을 때 `악셀을 밟아 가속한다` 라는 개념으로 생각해서 `acceleration` 이라고 표현했던 기억이 나네요 👍

dd-jiyun replied:
오! 저도 주노와 동일하게 떠올리긴 했어요! 
`acceleration`이 `power`보다 좀 더 직관적인 명칭인 것 같네요 👍🏻

Choi-JJunho commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,32 @@
+public class Car {
+    private static final int MIN_POWER_TO_MOVE = 3;
```
제가 읽기에는 `움직이기 위한 최소한의 힘`으로 해석되는데 그렇다면 3부터 자동차가 움직일 수 있는것일까? 라는 헷갈림이 생기네요
대표적인 예시로 테스트코드에서 동일한 이름의 상수는 4로 정의가 되어있네요!

이런 상황에서는 어떤 네이밍이 적절할까요? 🤔

dd-jiyun replied:
상수 네이밍을 많이 고민했는데 주노 말씀대로 헷갈릴 수 있을 것 같아요. 🥲
더 명확한 의미가 드러나도록 네이밍을 개선해보겠습니다!

그리고 수정하면서 하나 더 고민이 생겼어요.
현재는 Car와 CarTest 각각에 상수를 따로 선언해줬는데,
Car 클래스의 상수를 `public static`으로 열어두면 테스트에서도 동일한 상수를 사용할 수 있어서 중복을 줄일 수 있을 것 같더라고요!
하지만, 이 방식이 테스트가 구현에 의존하게 만드는 건 아닐까? 하는 생각도 듭니다.

이 부분에 대해서도 주노의 의견이 궁금합니다!

Choi-JJunho replied:
자동차의 속성으로써 사용하게된다면 큰 문제는 없지 않을까 생각이 드네요!

만약 더 복잡한 상황이 된다면 이러한 상수를 제공해주는 Factory 객체를 만드는것 꺼지 생각나지만.. 지금 구조로서는 과하가는 생각이드는군요 ㅎㅎ

Choi-JJunho commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,32 @@
+    private int position;
+
+    public Car(final String name) {
+        validateEmptyName(name);
```
validate 로직을 추가해주셨군요! 👍 
만약 이름에 대한 규칙이 복잡해지면 자동차 객체에 이름 검증 로직이 많이 생길 수 있을 것 같아요
이름 검증에 대한 책임을 분리하기 위해서는 어떤 방법을 사용해볼 수 있을까요?

hint: 원시값 포장

dd-jiyun replied:
오! 맞아요. 이름에 대한 유효성 검사가 복잡해질수록 Car 객체가 너무 많은 책임을 가지게 될 것 같아요.

말씀해주신 키워드와 같이 "이름"을 하나의 객체로 포장하면 Car 객체는 단순히 "이름을 가진다"는 책임만 가지게 할 수 있겠네요!

좋은 조언 감사합니다 😆

Choi-JJunho commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @ParameterizedTest
+    @NullAndEmptySource
```
👍👍

dd-jiyun committed:
refactor: 상수명 수정

- 자동차의 움직임을 위한 임계 값을 명확히 표현하도록 수정합니다.

dd-jiyun committed:
refactor: 랜덤값 생성 상수화

- 하드 코딩한 값을 명확하게 표현합니다.

dd-jiyun committed:
refactor: 이름을 원시값으로 포장하여 검증 책임 분리

- 이름 검증 로직을 `Name` 클래스에 포장하여, `Car` 객체의 책임을 줄이고, 이름 검증에 대한 변경 사항을 독립적으로 처리할 수 있도록 수정합니다.

dd-jiyun reviewed (commented):
좋은 리뷰 감사합니다! 👍🏻

Choi-JJunho reviewed (approved):
고생하셨습니다! 👍👍
다음 미션도 화이팅!
